File: ios/BetterLise/Features/Achievements/AchievementRules.swift
import Foundation

/// Progress figures for the achievements header.
struct AchievementSummary: Equatable, Sendable {
    let total: Int
    let unlocked: Int
    let legendary: Int
    let rare: Int
    let secrets: Int
    let secretsFound: Int

    init(_ achievements: [Achievement]) {
        let unlockedItems = achievements.filter(\.isUnlocked)
        total = achievements.count
        unlocked = unlockedItems.count
        legendary = unlockedItems.filter { $0.rarity == .legendary }.count
        rare = unlockedItems.filter { $0.rarity == .rare }.count
        secrets = achievements.filter(\.isSecret).count
        secretsFound = unlockedItems.filter(\.isSecret).count
    }

    /// Between 0 and 1.
    var progress: Double { total == 0 ? 0 : Double(unlocked) / Double(total) }
}

/// Decides which unlocks deserve a celebration, from the codes already celebrated on this device.
enum AchievementCelebration {
    /// First launch for an account: everything already unlocked counts as seen, except what this sync unlocked.
    static func initialSeen(achievements: [Achievement], newlyUnlocked: [String]) -> Set<String> {
        Set(achievements.filter(\.isUnlocked).map(\.code)).subtracting(newlyUnlocked)
    }

    /// Unlocked achievements not celebrated yet, in list order.
    static func pending(achievements: [Achievement], seen: Set<String>) -> [Achievement] {
        achievements.filter { $0.isUnlocked && !seen.contains($0.code) }
    }

    /// Reveal mode hides new grades: celebrating a 20/20 before its reveal would spoil it.
    static func shouldDefer(revealMode: Bool, unreadGrades: Int) -> Bool {
        revealMode && unreadGrades > 0
    }
}

/// SF Symbol for the icon keys sent by the API.
enum AchievementSymbol {
    static func name(for icon: String?) -> String {
        switch icon {
        case "rocket": "sparkles"
        case "trophy": "trophy.fill"
        case "fall": "chart.line.downtrend.xyaxis"
        case "reload": "arrow.counterclockwise"
        case "fire": "flame.fill"
        case "aim": "scope"
        case "heart": "heart.fill"
        case "experiment": "testtube.2"
        case "flag": "flag.fill"
        default: "star.fill"
        }
    }
}

File: android/app/src/main/java/com/betterlise/app/domain/AchievementRules.kt
package com.betterlise.app.domain

import com.betterlise.app.data.api.Achievement
import com.betterlise.app.data.api.AchievementRarity

/** Progress figures for the achievements header. */
data class AchievementSummary(
    val total: Int,
    val unlocked: Int,
    val legendary: Int,
    val rare: Int,
    val secrets: Int,
    val secretsFound: Int,
) {
    /** Between 0 and 1. */
    val progress: Double get() = if (total == 0) 0.0 else unlocked.toDouble() / total

    companion object {
        fun of(achievements: List<Achievement>): AchievementSummary {
            val unlocked = achievements.filter { it.isUnlocked }
            return AchievementSummary(
                total = achievements.size,
                unlocked = unlocked.size,
                legendary = unlocked.count { it.rarity == AchievementRarity.Legendary },
                rare = unlocked.count { it.rarity == AchievementRarity.Rare },
                secrets = achievements.count { it.isSecret },
                secretsFound = unlocked.count { it.isSecret },
            )
        }
    }
}

/** Decides which unlocks deserve a celebration, from the codes already celebrated on this device. */
object AchievementCelebration {
    /** First launch for an account: everything already unlocked counts as seen, except what this sync unlocked. */
    fun initialSeen(achievements: List<Achievement>, newlyUnlocked: List<String>): Set<String> =
        achievements.filter { it.isUnlocked }.map { it.code }.toSet() - newlyUnlocked.toSet()

    /** Unlocked achievements not celebrated yet, in list order. */
    fun pending(achievements: List<Achievement>, seen: Set<String>): List<Achievement> =
        achievements.filter { it.isUnlocked && it.code !in seen }

    /** Reveal mode hides new grades: celebrating a 20/20 before its reveal would spoil it. */
    fun shouldDefer(revealMode: Boolean, unreadGrades: Int): Boolean = revealMode && unreadGrades > 0
}
